### PR TITLE
Adding gradle publish task to ppl & protocol module to create slim jars

### PR DESCRIPTION
**Pull request by zane-neo**
### Description
Adding gradle publish task to ppl & protocol module to create slim jars
 
### Issues Resolved
Adding gradle publish task to ppl & protocol module to create slim jars
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dai-chen — COMMENTED** [review]
PPL transport model class is in `plugin` module. So how we add this dependency? Just thinking if we have to depend on the current SQL artifact https://mvnrepository.com/artifact/org.opensearch.plugin/opensearch-sql-plugin anyway, probably we don't need this slim jar?

File: protocol/build.gradle [hunk 1]
```diff
@@ -89,3 +90,39 @@ jacocoTestCoverageVerification {
 }
 check.dependsOn jacocoTestCoverageVerification
 jacocoTestCoverageVerification.dependsOn jacocoTestReport
+
+publishing {
+    publications {
+        mavenJava(MavenPublication) {
+            groupId = 'org.opensearch'
+            artifactId = 'opensearch-protocol'
```
[the remark this answers is not in the document]
Comment (zane-neo) [reply protocol/build.gradle:98]: This won't change the plugin's location in maven repo, this only controls the slim jar's location and it seems opensearch jars are always under org.opensearch folder: https://mvnrepository.com/search?q=org.opensearch&p=1

**Comment by zane-neo** [comment]
> PPL transport model class is in `plugin` module. So how we add this dependency? Just thinking if we have to depend on the current SQL artifact https://mvnrepository.com/artifact/org.opensearch.plugin/opensearch-sql-plugin anyway, probably we don't need this slim jar?

We still need these slim jars since when building ml-common plugin(which is a shadow jar), it'll recursively find all dependent jars to its classpath, without these jars, it's not possible to build a ml-common plugin successfully. Also, in runtime, different plugins are loaded by different classloaders, so ml-commons can not see the classes under sql plugin, without these jars under ml-commons plugin, there'll be runtime errors.

**Comment by dai-chen** [comment]
> > PPL transport model class is in `plugin` module. So how we add this dependency? Just thinking if we have to depend on the current SQL artifact https://mvnrepository.com/artifact/org.opensearch.plugin/opensearch-sql-plugin anyway, probably we don't need this slim jar?
> 
> We still need these slim jars since when building ml-common plugin(which is a shadow jar), it'll recursively find all dependent jars to its classpath, without these jars, it's not possible to build a ml-common plugin successfully. Also, in runtime, different plugins are loaded by different classloaders, so ml-commons can not see the classes under sql plugin, without these jars under ml-commons plugin, there'll be runtime errors.

I downloaded the current SQL zip artifact published to Maven to double check. It actually contains all jar files, including plugin, PPL and protocol module jar. I was thinking any chance your gradle depends on it directly and remove those you feel not needed at runtime? Let me know if I missed anything.

```
opensearch-sql-plugin-2.11.1.0 ll
total 105888
-rw-r--r--@ 1 daichen  staff    11K Feb  1  1980 LICENSE.txt
-rw-r--r--@ 1 daichen  staff   721B Feb  1  1980 NOTICE.txt
-rw-r--r--@ 1 daichen  staff   234K Feb  1  1980 ST4-4.0.8.jar
-rw-r--r--@ 1 daichen  staff    17K Feb  1  1980 annotations-13.0.jar
-rw-r--r--@ 1 daichen  staff   164K Feb  1  1980 antlr-runtime-3.5.2.jar
-rw-r--r--@ 1 daichen  staff   1.3M Feb  1  1980 antlr4-4.7.1.jar
-rw-r--r--@ 1 daichen  staff   329K Feb  1  1980 antlr4-runtime-4.7.1.jar
-rw-r--r--@ 1 daichen  staff   231K Feb  1  1980 aws-encryption-sdk-java-2.4.1.jar
-rw-r--r--@ 1 daichen  staff   1.0M Feb  1  1980 aws-java-sdk-core-1.12.545.jar
-rw-r--r--@ 1 daichen  staff   1.2M Feb  1  1980 aws-java-sdk-emr-1.12.545.jar
-rw-r--r--@ 1 daichen  staff   312K Feb  1  1980 aws-java-sdk-emrserverless-1.12.545.jar
-rw-r--r--@ 1 daichen  staff   151K Feb  1  1980 aws-java-sdk-sts-1.12.545.jar
-rw-r--r--@ 1 daichen  staff   8.0M Feb  1  1980 bcprov-ext-jdk18on-1.75.jar
-rw-r--r--@ 1 daichen  staff   219K Feb  1  1980 checker-qual-3.33.0.jar
-rw-r--r--@ 1 daichen  staff    46K Feb  1  1980 common-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   700K Feb  1  1980 common-utils-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   241K Feb  1  1980 commons-beanutils-1.9.4.jar
-rw-r--r--@ 1 daichen  staff   336K Feb  1  1980 commons-codec-1.13.jar
-rw-r--r--@ 1 daichen  staff   575K Feb  1  1980 commons-collections-3.2.2.jar
-rw-r--r--@ 1 daichen  staff   192K Feb  1  1980 commons-digester-2.1.jar
-rw-r--r--@ 1 daichen  staff   279K Feb  1  1980 commons-io-2.8.0.jar
-rw-r--r--@ 1 daichen  staff   574K Feb  1  1980 commons-lang3-3.12.0.jar
-rw-r--r--@ 1 daichen  staff    60K Feb  1  1980 commons-logging-1.2.jar
-rw-r--r--@ 1 daichen  staff   2.1M Feb  1  1980 commons-math3-3.6.1.jar
-rw-r--r--@ 1 daichen  staff   185K Feb  1  1980 commons-validator-1.7.jar
-rw-r--r--@ 1 daichen  staff   659K Feb  1  1980 core-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff    60K Feb  1  1980 datasources-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   1.9M Feb  1  1980 druid-1.0.15.jar
-rw-r--r--@ 1 daichen  staff    16K Feb  1  1980 error_prone_annotations-2.18.0.jar
-rw-r--r--@ 1 daichen  staff   4.5K Feb  1  1980 failureaccess-1.0.1.jar
-rw-r--r--@ 1 daichen  staff    93K Feb  1  1980 geo-2.11.1.jar
-rw-r--r--@ 1 daichen  staff   252K Feb  1  1980 gson-2.8.9.jar
-rw-r--r--@ 1 daichen  staff   2.9M Feb  1  1980 guava-32.0.1-jre.jar
-rw-r--r--@ 1 daichen  staff   177K Feb  1  1980 httpasyncclient-4.1.5.jar
-rw-r--r--@ 1 daichen  staff   767K Feb  1  1980 httpclient-4.5.14.jar
-rw-r--r--@ 1 daichen  staff   321K Feb  1  1980 httpcore-4.4.15.jar
-rw-r--r--@ 1 daichen  staff   360K Feb  1  1980 httpcore-nio-4.4.16.jar
-rw-r--r--@ 1 daichen  staff    11M Feb  1  1980 icu4j-58.2.jar
-rw-r--r--@ 1 daichen  staff   552K Feb  1  1980 ion-java-1.0.2.jar
-rw-r--r--@ 1 daichen  staff   869K Feb  1  1980 ipaddress-5.4.0.jar
-rw-r--r--@ 1 daichen  staff   9.1K Feb  1  1980 j2objc-annotations-2.8.jar
-rw-r--r--@ 1 daichen  staff    74K Feb  1  1980 jackson-annotations-2.15.2.jar
-rw-r--r--@ 1 daichen  staff   1.5M Feb  1  1980 jackson-databind-2.15.2.jar
-rw-r--r--@ 1 daichen  staff   764K Feb  1  1980 javassist-3.26.0-GA.jar
-rw-r--r--@ 1 daichen  staff    83K Feb  1  1980 javax.json-1.0.4.jar
-rw-r--r--@ 1 daichen  staff    27K Feb  1  1980 jmespath-java-1.12.545.jar
-rw-r--r--@ 1 daichen  staff    71K Feb  1  1980 json-20230227.jar
-rw-r--r--@ 1 daichen  staff    19K Feb  1  1980 jsr305-3.0.2.jar
-rw-r--r--@ 1 daichen  staff   2.8M Feb  1  1980 kotlin-reflect-1.4.30.jar
-rw-r--r--@ 1 daichen  staff   1.6M Feb  1  1980 kotlin-stdlib-1.9.0.jar
-rw-r--r--@ 1 daichen  staff   196K Feb  1  1980 kotlin-stdlib-common-1.6.0.jar
-rw-r--r--@ 1 daichen  staff   958B Feb  1  1980 kotlin-stdlib-jdk7-1.9.0.jar
-rw-r--r--@ 1 daichen  staff   968B Feb  1  1980 kotlin-stdlib-jdk8-1.8.0.jar
-rw-r--r--@ 1 daichen  staff   951K Feb  1  1980 legacy-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   2.1K Feb  1  1980 listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
-rw-r--r--@ 1 daichen  staff   764K Feb  1  1980 okhttp-4.10.0.jar
-rw-r--r--@ 1 daichen  staff    14K Feb  1  1980 okhttp-aws-signer-1.0.2.jar
-rw-r--r--@ 1 daichen  staff   352K Feb  1  1980 okio-jvm-3.4.0.jar
-rw-r--r--@ 1 daichen  staff   248K Feb  1  1980 opensearch-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   499K Feb  1  1980 opensearch-ml-client-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff    67K Feb  1  1980 opensearch-rest-client-2.11.1.jar
-rw-r--r--@ 1 daichen  staff    37K Feb  1  1980 opensearch-sql-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff    66K Feb  1  1980 opensearch-ssl-config-2.11.1.jar
-rw-r--r--@ 1 daichen  staff    26K Feb  1  1980 org.abego.treelayout.core-1.0.3.jar
-rw-r--r--@ 1 daichen  staff    85K Feb  1  1980 parent-join-client-2.11.1.jar
-rw-r--r--@ 1 daichen  staff   1.8K Feb  1  1980 plugin-descriptor.properties
-rw-r--r--@ 1 daichen  staff   1.0K Feb  1  1980 plugin-security.policy
-rw-r--r--@ 1 daichen  staff   334K Feb  1  1980 ppl-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff    24K Feb  1  1980 presto-matching-0.240.jar
-rw-r--r--@ 1 daichen  staff    87K Feb  1  1980 prometheus-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff    38K Feb  1  1980 protocol-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   103K Feb  1  1980 reflections-0.9.12.jar
-rw-r--r--@ 1 daichen  staff   117K Feb  1  1980 reindex-client-2.11.1.jar
-rw-r--r--@ 1 daichen  staff    52K Feb  1  1980 resilience4j-core-1.5.0.jar
-rw-r--r--@ 1 daichen  staff    45K Feb  1  1980 resilience4j-retry-1.5.0.jar
-rw-r--r--@ 1 daichen  staff    40K Feb  1  1980 slf4j-api-1.7.36.jar
-rw-r--r--@ 1 daichen  staff   944K Feb  1  1980 spark-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   360K Feb  1  1980 sql-2.11.1.0.jar
-rw-r--r--@ 1 daichen  staff   877K Feb  1  1980 vavr-0.10.2.jar
-rw-r--r--@ 1 daichen  staff   3.0K Feb  1  1980 vavr-match-0.10.2.jar
```

**Comment by zane-neo** [comment]
> > > PPL transport model class is in `plugin` module. So how we add this dependency? Just thinking if we have to depend on the current SQL artifact https://mvnrepository.com/artifact/org.opensearch.plugin/opensearch-sql-plugin anyway, probably we don't need this slim jar?
> > 
> > 
> > We still need these slim jars since when building ml-common plugin(which is a shadow jar), it'll recursively find all dependent jars to its classpath, without these jars, it's not possible to build a ml-common plugin successfully. Also, in runtime, different plugins are loaded by different classloaders, so ml-commons can not see the classes under sql plugin, without these jars under ml-commons plugin, there'll be runtime errors.
> 
> I downloaded the current SQL zip artifact published to Maven to double check. It actually contains all jar files, including plugin, PPL and protocol module jar. I was thinking any chance your gradle depends on it directly and remove those you feel not needed at runtime? Let me know if I missed anything.
> 
> ```
> opensearch-sql-plugin-2.11.1.0 ll
> total 105888
> -rw-r--r--@ 1 daichen  staff    11K Feb  1  1980 LICENSE.txt
> -rw-r--r--@ 1 daichen  staff   721B Feb  1  1980 NOTICE.txt
> -rw-r--r--@ 1 daichen  staff   234K Feb  1  1980 ST4-4.0.8.jar
> -rw-r--r--@ 1 daichen  staff    17K Feb  1  1980 annotations-13.0.jar
> -rw-r--r--@ 1 daichen  staff   164K Feb  1  1980 antlr-runtime-3.5.2.jar
> -rw-r--r--@ 1 daichen  staff   1.3M Feb  1  1980 antlr4-4.7.1.jar
> -rw-r--r--@ 1 daichen  staff   329K Feb  1  1980 antlr4-runtime-4.7.1.jar
> -rw-r--r--@ 1 daichen  staff   231K Feb  1  1980 aws-encryption-sdk-java-2.4.1.jar
> -rw-r--r--@ 1 daichen  staff   1.0M Feb  1  1980 aws-java-sdk-core-1.12.545.jar
> -rw-r--r--@ 1 daichen  staff   1.2M Feb  1  1980 aws-java-sdk-emr-1.12.545.jar
> -rw-r--r--@ 1 daichen  staff   312K Feb  1  1980 aws-java-sdk-emrserverless-1.12.545.jar
> -rw-r--r--@ 1 daichen  staff   151K Feb  1  1980 aws-java-sdk-sts-1.12.545.jar
> -rw-r--r--@ 1 daichen  staff   8.0M Feb  1  1980 bcprov-ext-jdk18on-1.75.jar
> -rw-r--r--@ 1 daichen  staff   219K Feb  1  1980 checker-qual-3.33.0.jar
> -rw-r--r--@ 1 daichen  staff    46K Feb  1  1980 common-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   700K Feb  1  1980 common-utils-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   241K Feb  1  1980 commons-beanutils-1.9.4.jar
> -rw-r--r--@ 1 daichen  staff   336K Feb  1  1980 commons-codec-1.13.jar
> -rw-r--r--@ 1 daichen  staff   575K Feb  1  1980 commons-collections-3.2.2.jar
> -rw-r--r--@ 1 daichen  staff   192K Feb  1  1980 commons-digester-2.1.jar
> -rw-r--r--@ 1 daichen  staff   279K Feb  1  1980 commons-io-2.8.0.jar
> -rw-r--r--@ 1 daichen  staff   574K Feb  1  1980 commons-lang3-3.12.0.jar
> -rw-r--r--@ 1 daichen  staff    60K Feb  1  1980 commons-logging-1.2.jar
> -rw-r--r--@ 1 daichen  staff   2.1M Feb  1  1980 commons-math3-3.6.1.jar
> -rw-r--r--@ 1 daichen  staff   185K Feb  1  1980 commons-validator-1.7.jar
> -rw-r--r--@ 1 daichen  staff   659K Feb  1  1980 core-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff    60K Feb  1  1980 datasources-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   1.9M Feb  1  1980 druid-1.0.15.jar
> -rw-r--r--@ 1 daichen  staff    16K Feb  1  1980 error_prone_annotations-2.18.0.jar
> -rw-r--r--@ 1 daichen  staff   4.5K Feb  1  1980 failureaccess-1.0.1.jar
> -rw-r--r--@ 1 daichen  staff    93K Feb  1  1980 geo-2.11.1.jar
> -rw-r--r--@ 1 daichen  staff   252K Feb  1  1980 gson-2.8.9.jar
> -rw-r--r--@ 1 daichen  staff   2.9M Feb  1  1980 guava-32.0.1-jre.jar
> -rw-r--r--@ 1 daichen  staff   177K Feb  1  1980 httpasyncclient-4.1.5.jar
> -rw-r--r--@ 1 daichen  staff   767K Feb  1  1980 httpclient-4.5.14.jar
> -rw-r--r--@ 1 daichen  staff   321K Feb  1  1980 httpcore-4.4.15.jar
> -rw-r--r--@ 1 daichen  staff   360K Feb  1  1980 httpcore-nio-4.4.16.jar
> -rw-r--r--@ 1 daichen  staff    11M Feb  1  1980 icu4j-58.2.jar
> -rw-r--r--@ 1 daichen  staff   552K Feb  1  1980 ion-java-1.0.2.jar
> -rw-r--r--@ 1 daichen  staff   869K Feb  1  1980 ipaddress-5.4.0.jar
> -rw-r--r--@ 1 daichen  staff   9.1K Feb  1  1980 j2objc-annotations-2.8.jar
> -rw-r--r--@ 1 daichen  staff    74K Feb  1  1980 jackson-annotations-2.15.2.jar
> -rw-r--r--@ 1 daichen  staff   1.5M Feb  1  1980 jackson-databind-2.15.2.jar
> -rw-r--r--@ 1 daichen  staff   764K Feb  1  1980 javassist-3.26.0-GA.jar
> -rw-r--r--@ 1 daichen  staff    83K Feb  1  1980 javax.json-1.0.4.jar
> -rw-r--r--@ 1 daichen  staff    27K Feb  1  1980 jmespath-java-1.12.545.jar
> -rw-r--r--@ 1 daichen  staff    71K Feb  1  1980 json-20230227.jar
> -rw-r--r--@ 1 daichen  staff    19K Feb  1  1980 jsr305-3.0.2.jar
> -rw-r--r--@ 1 daichen  staff   2.8M Feb  1  1980 kotlin-reflect-1.4.30.jar
> -rw-r--r--@ 1 daichen  staff   1.6M Feb  1  1980 kotlin-stdlib-1.9.0.jar
> -rw-r--r--@ 1 daichen  staff   196K Feb  1  1980 kotlin-stdlib-common-1.6.0.jar
> -rw-r--r--@ 1 daichen  staff   958B Feb  1  1980 kotlin-stdlib-jdk7-1.9.0.jar
> -rw-r--r--@ 1 daichen  staff   968B Feb  1  1980 kotlin-stdlib-jdk8-1.8.0.jar
> -rw-r--r--@ 1 daichen  staff   951K Feb  1  1980 legacy-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   2.1K Feb  1  1980 listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
> -rw-r--r--@ 1 daichen  staff   764K Feb  1  1980 okhttp-4.10.0.jar
> -rw-r--r--@ 1 daichen  staff    14K Feb  1  1980 okhttp-aws-signer-1.0.2.jar
> -rw-r--r--@ 1 daichen  staff   352K Feb  1  1980 okio-jvm-3.4.0.jar
> -rw-r--r--@ 1 daichen  staff   248K Feb  1  1980 opensearch-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   499K Feb  1  1980 opensearch-ml-client-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff    67K Feb  1  1980 opensearch-rest-client-2.11.1.jar
> -rw-r--r--@ 1 daichen  staff    37K Feb  1  1980 opensearch-sql-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff    66K Feb  1  1980 opensearch-ssl-config-2.11.1.jar
> -rw-r--r--@ 1 daichen  staff    26K Feb  1  1980 org.abego.treelayout.core-1.0.3.jar
> -rw-r--r--@ 1 daichen  staff    85K Feb  1  1980 parent-join-client-2.11.1.jar
> -rw-r--r--@ 1 daichen  staff   1.8K Feb  1  1980 plugin-descriptor.properties
> -rw-r--r--@ 1 daichen  staff   1.0K Feb  1  1980 plugin-security.policy
> -rw-r--r--@ 1 daichen  staff   334K Feb  1  1980 ppl-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff    24K Feb  1  1980 presto-matching-0.240.jar
> -rw-r--r--@ 1 daichen  staff    87K Feb  1  1980 prometheus-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff    38K Feb  1  1980 protocol-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   103K Feb  1  1980 reflections-0.9.12.jar
> -rw-r--r--@ 1 daichen  staff   117K Feb  1  1980 reindex-client-2.11.1.jar
> -rw-r--r--@ 1 daichen  staff    52K Feb  1  1980 resilience4j-core-1.5.0.jar
> -rw-r--r--@ 1 daichen  staff    45K Feb  1  1980 resilience4j-retry-1.5.0.jar
> -rw-r--r--@ 1 daichen  staff    40K Feb  1  1980 slf4j-api-1.7.36.jar
> -rw-r--r--@ 1 daichen  staff   944K Feb  1  1980 spark-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   360K Feb  1  1980 sql-2.11.1.0.jar
> -rw-r--r--@ 1 daichen  staff   877K Feb  1  1980 vavr-0.10.2.jar
> -rw-r--r--@ 1 daichen  staff   3.0K Feb  1  1980 vavr-match-0.10.2.jar
> ```

We do have another option which is to create a fat/shadow jar of plugin, we tried this option and have a POC PR here: https://github.com/Hailong-am/sql/commit/b7635caeb2223ef5681d7262fe05e16a1c1d5c84. But this could have impact on publishing plugin zip. Publishing a slim jar is always better than shadow jar which can avoid a lot of jar conflict, so we prefer to publish slim jars like in this PR.

**Comment by zane-neo** [comment]
Using zip extraction approach to create a fat jar instead of publish slim jars.

**Pull request closed without merging**